Gems/Atom/RHI/Code/Source/RHI/MultiDeviceShaderResourceGroupPool.cpp

```diff
@@ -0,0 +1,197 @@
+                return ResultCode::InvalidArgument;
+            }
+        }
+
+        ResultCode resultCode = MultiDeviceResourcePool::Init(
+            deviceMask,
+            [this, &descriptor]()
+            {
+                IterateDevices(
+                    [this, &descriptor](int deviceIndex)
+                    {
+                        auto* device = RHISystemInterface::Get()->GetDevice(deviceIndex);
+
+                        m_deviceObjects[deviceIndex] = Factory::Get().CreateShaderResourceGroupPool();
+                        GetDeviceShaderResourceGroupPool(deviceIndex)->Init(*device, descriptor);
+
+                        return true;
+                    });
+                return ResultCode::Success;
+            });
+
+        if (resultCode != ResultCode::Success)
+        {
+            // Reset already initialized device-specific ShaderResourceGroupPools and set deviceMask to 0
+            m_deviceObjects.clear();
+            MultiDeviceObject::Init(static_cast<MultiDevice::DeviceMask>(0u));
+            return resultCode;
+        }
+
+        m_descriptor = descriptor;
+        m_hasBufferGroup = m_descriptor.m_layout->GetGroupSizeForBuffers() > 0;
+        m_hasImageGroup = m_descriptor.m_layout->GetGroupSizeForImages() > 0;
+        m_hasSamplerGroup = m_descriptor.m_layout->GetGroupSizeForSamplers() > 0;
+        m_hasConstants = m_descriptor.m_layout->GetConstantDataSize() > 0;
+
+        return ResultCode::Success;
+    }
+
+    ResultCode MultiDeviceShaderResourceGroupPool::InitGroup(MultiDeviceShaderResourceGroup& group)
+    {
+        ResultCode resultCode = MultiDeviceResourcePool::InitResource(
+            &group,
+            [this, &group]()
+            {
+                return IterateObjects<ShaderResourceGroupPool>([this, &group](auto deviceIndex, [[maybe_unused]] auto deviceShaderResourceGroupPool)
+                {
+                    group.m_deviceObjects[deviceIndex] = Factory::Get().CreateShaderResourceGroup();
+                    return GetDeviceShaderResourceGroupPool(deviceIndex)->InitGroup(*group.GetDeviceShaderResourceGroup(deviceIndex));
+                });
+            });
+
+        if (resultCode == ResultCode::Success)
+        {
+            const ShaderResourceGroupLayout* layout = GetLayout();
+
+            // Pre-initialize the data so that we can build view diffs later.
+            group.m_mdData = MultiDeviceShaderResourceGroupData(GetDeviceMask(), layout);
+
+            // Cache off the binding slot for one less indirection.
+            group.m_bindingSlot = layout->GetBindingSlot();
+        }
+        else
+        {
+            // Reset already initialized device-specific ShaderResourceGroupPools and set deviceMask to 0
+            m_deviceObjects.clear();
+            MultiDeviceObject::Init(static_cast<MultiDevice::DeviceMask>(0u));
```

Comment:
Our reasoning is that it is not clearly defined what to do and what to report if one or multiple initializations fail. 
Hence, our current strategy is to reset the MD object, clearing all SD objects and reporting an error back. 

group.m_deviceObjects[deviceIndex] does not have to be reset, since it was not initialized, but the question is what to do with already initialized groups. 

Do you see a better way?